MetaTensorDesc changes for reconstructing proper FakeTensors

aorenste opened the pull request:
A few changes to MetaTensorDesc and friends:

1. Change view_func from a raw method to an ADT where the common case (FakeTensor._view_func_unsafe) is a simple representation instead.
2. (minor) Remove and fix some `type: ignore`s added by #141839
3. (minor) Fix _UNSERIALIZABLE to be a set instead of a dict which is converted into a set each time it's used.
 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #142072
* #141659
* __->__ #141926

aorenste committed:
MetaTensorDesc changes for reconstructing proper FakeTensors

[ghstack-poisoned]

aorenste committed:
Update on "MetaTensorDesc changes for reconstructing proper FakeTensors"

A few changes to MetaTensorDesc and friends:

1. Change view_func from a raw method to an ADT where the common case (FakeTensor._view_func_unsafe) is a simple representation instead.
2. Allow passing an "override_device" to meta_tensor() (or more importantly allow you to NOT override if you don't want to).  The existing behavior forced the device to "meta" - but for reconstructing FakeTensor that's not the behavior we want.
3. (minor) Fix _UNSERIALIZABLE to be a set instead of a dict which is converted into a set each time it's used.
 




[ghstack-poisoned]

aorenste commented on torch/_subclasses/meta_utils.py:
```diff
@@ -345,9 +356,11 @@ def describe_tensor(
+
         # TODO: Is it important to enable torch.inference_mode before querying
         # these values?
-        r = MetaTensorDesc(
```
mypy is being weird and requiring a type decl here...

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -476,6 +487,64 @@ def as_json(self, describer_id: _DescriberId) -> Dict[str, object]:
 
 
+@dataclass(frozen=True)
+class _ViewFunc(Generic[_TensorT]):
```
FWIW I don't really see why this should be private, if MetaTensorDesc isn't considered private per this file.

aorenste replied:
It doesn't have to be - but it's kind of an implementation detail...

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -1582,16 +1643,14 @@ def is_c_of_r(
                                 strides,
                                 dtype=t.dtype,
-                                device="meta",
+                                device=override_device or t.device,
```
I can't tell if this is semantics preserving or not

aorenste replied:
This is on the older version of this PR - there was a collision w/ #141839 (which landed yesterday) which implements similar functionality in a different way - so I ripped the override stuff out and just used that.

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -1809,6 +1865,7 @@ def __call__(
                 callback_,
                 source,
                 symbolic_context,
+                override_device="meta",
```
oh I guess it is

ezyang commented:
I think you need to recursively propagate the force device thing. But I am also not too sure about what that is about

aorenste committed:
Update on "MetaTensorDesc changes for reconstructing proper FakeTensors"

A few changes to MetaTensorDesc and friends:

1. Change view_func from a raw method to an ADT where the common case (FakeTensor._view_func_unsafe) is a simple representation instead.
2. (minor) Remove and fix some `type: ignore`s added by #141839
3. (minor) Fix _UNSERIALIZABLE to be a set instead of a dict which is converted into a set each time it's used.
 




[ghstack-poisoned]

aorenste committed:
Update on "MetaTensorDesc changes for reconstructing proper FakeTensors"

A few changes to MetaTensorDesc and friends:

1. Change view_func from a raw method to an ADT where the common case (FakeTensor._view_func_unsafe) is a simple representation instead.
2. (minor) Remove and fix some `type: ignore`s added by #141839
3. (minor) Fix _UNSERIALIZABLE to be a set instead of a dict which is converted into a set each time it's used.
 




[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
